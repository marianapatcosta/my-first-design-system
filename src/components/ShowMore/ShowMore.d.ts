import React, { CSSProperties, ReactNode } from 'react';
export interface ShowMoreProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * show more content
     */
    content: string | ReactNode;
    /**
     * number of content lines that are visible when show more is collapsed
     */
    numberOfPreviewLines?: number;
}
export declare const ShowMore: React.FC<ShowMoreProps>;
