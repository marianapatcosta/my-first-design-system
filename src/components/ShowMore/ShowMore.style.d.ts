import { MouseEventHandler } from 'react';
export declare const StyledShowMore: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const StyledShowMoreLink: import("styled-components").StyledComponent<"a", any, {
    onClick?: MouseEventHandler<HTMLAnchorElement> | undefined;
}, never>;
export declare const StyledShowMoreIcon: import("styled-components").StyledComponent<"img", any, {
    showMore?: boolean | undefined;
}, never>;
export declare const ShowMoreContent: import("styled-components").StyledComponent<"div", any, {
    showMore?: boolean | undefined;
}, never>;
