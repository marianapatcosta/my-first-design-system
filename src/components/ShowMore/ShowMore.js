var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, } from 'react';
import { DoubleChevron } from '../../assets/icons';
import { StyledShowMore, ShowMoreContent, StyledShowMoreLink, StyledShowMoreIcon, } from './ShowMore.style';
export var ShowMore = function (_a) {
    var content = _a.content, numberOfPreviewLines = _a.numberOfPreviewLines, props = __rest(_a, ["content", "numberOfPreviewLines"]);
    var _b = useState(false), showMore = _b[0], setShowMore = _b[1];
    var toggleShowMore = function (event) {
        event.preventDefault();
        setShowMore(function (prevShowMore) { return !prevShowMore; });
    };
    var label = showMore ? 'show less' : 'show more';
    var lineHeight = 1.2; // in rem
    return (_jsxs(StyledShowMore, __assign({}, props, { children: [_jsx(ShowMoreContent, __assign({ showMore: showMore, style: {
                    maxHeight: "" + (!showMore
                        ? (numberOfPreviewLines &&
                            numberOfPreviewLines * lineHeight - 0.4) + "rem"
                        : 'inherit'),
                } }, { children: content }), void 0),
            _jsxs(StyledShowMoreLink, __assign({ "data-testid": 'show-more-link', onClick: toggleShowMore, href: '' }, { children: [label, ' ', _jsx(StyledShowMoreIcon, { showMore: showMore, src: DoubleChevron, alt: label + " icon" }, void 0)] }), void 0)] }), void 0));
};
ShowMore.defaultProps = {
    id: '',
    className: '',
    style: {},
    content: '',
    numberOfPreviewLines: 5,
};
