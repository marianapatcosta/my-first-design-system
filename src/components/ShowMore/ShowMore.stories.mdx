import { Meta, Story, ArgsTable, Canvas, Preview } from '@storybook/addon-docs/blocks';
import { ShowMore } from "./ShowMore";

<Meta title="Design System/ShowMore" component={ShowMore} />

# ShowMore
<br/><br/> 

### Displays partial content by default and, on demand, displays all the content

Use ShowMore to present a truncated text /HTML element list that can be fully displayed if `show more` link is clicked

<Preview withToolbar>
  <Story name="Standard">
    <ShowMore content="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec vitae mollis velit. Vestibulum lacinia nisi vitae lorem sollicitudin, vel ultricies enim suscipit. Etiam vel sodales dolor. Nunc iaculis quis arcu faucibus tincidunt. Maecenas blandit dapibus porttitor. Quisque sed mauris elit. Ut placerat, augue ut blandit ultricies, nunc nulla vehicula mi, in vehicula urna lectus nec nibh. Aenean quis risus imperdiet, bibendum mauris vitae, consequat tortor. Donec aliquam ante vel turpis fringilla facilisis. Integer varius non lacus id ultrices. In eleifend dolor ante, congue placerat elit maximus eget. Duis varius ligula erat, et varius neque euismod ac. In vitae posuere nisl. Sed non justo tortor. Integer risus est, feugiat at ultrices vitae, pretium in est. In vel tristique augue. Nam tempor neque erat, vitae cursus diam eleifend non. Aliquam libero odio, vestibulum at nunc sit amet, suscipit suscipit libero. Nam ac pretium felis, eu convallis libero. Phasellus quis odio sit amet nibh tempor placerat non sed nunc. Aliquam pellentesque efficitur iaculis. Cras eget gravida leo.  Ut sed mattis erat. In in nunc eu ligula egestas viverra. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Curabitur fermentum sem mauris, semper consequat urna sagittis sed. Maecenas lacinia nulla lorem, quis condimentum orci ultricies quis. Aliquam erat volutpat. Vestibulum volutpat, metus quis malesuada lobortis, quam erat accumsan lectus, nec bibendum leo libero facilisis risus. Nulla sed viverra orci." />
  </Story>
</Preview>

<ArgsTable of={ShowMore} />

### ShowMore With list:
  <Story name="WithList">
    <ShowMore
      content={
        <ul>
          {["item 1", "item 2", "item 3", "item 4", "item 5", "item 6"].map(
            (item, index) => (
              <li key={`example-list-${index}`}>{item}</li>
            )
          )}
        </ul>
      }
    />
  </Story>

### Playground

Experiment with ShowMore Component with the Controls addon in the Canvas tab.

export const Template = (args) => <ShowMore {...args} />

<Canvas>
  <Story name="Controls" args={{
    content: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec vitae mollis velit. Vestibulum lacinia nisi vitae lorem sollicitudin, vel ultricies enim suscipit. Etiam vel sodales dolor. Nunc iaculis quis arcu faucibus tincidunt. Maecenas blandit dapibus porttitor. Quisque sed mauris elit. Ut placerat, augue ut blandit ultricies, nunc nulla vehicula mi, in vehicula urna lectus nec nibh. Aenean quis risus imperdiet, bibendum mauris vitae, consequat tortor. Donec aliquam ante vel turpis fringilla facilisis. Integer varius non lacus id ultrices. In eleifend dolor ante, congue placerat elit maximus eget. Duis varius ligula erat, et varius neque euismod ac. In vitae posuere nisl. Sed non justo tortor. Integer risus est, feugiat at ultrices vitae, pretium in est. In vel tristique augue. Nam tempor neque erat, vitae cursus diam eleifend non. Aliquam libero odio, vestibulum at nunc sit amet, suscipit suscipit libero. Nam ac pretium felis, eu convallis libero. Phasellus quis odio sit amet nibh tempor placerat non sed nunc. Aliquam pellentesque efficitur iaculis. Cras eget gravida leo.  Ut sed mattis erat. In in nunc eu ligula egestas viverra. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Curabitur fermentum sem mauris, semper consequat urna sagittis sed. Maecenas lacinia nulla lorem, quis condimentum orci ultricies quis. Aliquam erat volutpat. Vestibulum volutpat, metus quis malesuada lobortis, quam erat accumsan lectus, nec bibendum leo libero facilisis risus. Nulla sed viverra orci.`,
  }}>
    {Template.bind({})}
  </Story>
</Canvas> 
