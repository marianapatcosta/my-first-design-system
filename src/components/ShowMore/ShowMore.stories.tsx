import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { ShowMore, ShowMoreProps } from './ShowMore';

export default {
  title: 'ShowMore',
  component: ShowMore,
  args: {},
} as Meta;

const Template: Story<ShowMoreProps> = (args) =>
  <ShowMore {...args} />;

  export const Default = Template.bind({});
  Default.args = {

  };