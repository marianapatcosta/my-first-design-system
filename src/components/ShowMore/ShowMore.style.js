var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';
export var StyledShowMore = styled.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ", "\n"], ["\n  ", "\n"])), baseStyles);
export var StyledShowMoreLink = styled.a(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  text-decoration: none;\n  color: ", ";\n  font-size: 90%;\n  font-weight: 700;\n  margin: 1rem 0;\n  float: right;\n  cursor: pointer;\n"], ["\n  text-decoration: none;\n  color: ", ";\n  font-size: 90%;\n  font-weight: 700;\n  margin: 1rem 0;\n  float: right;\n  cursor: pointer;\n"])), colors.highlight);
export var StyledShowMoreIcon = styled.img(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", ";\n  transition: transform 0.5s ease-in-out;\n  margin-left: 0.3rem;\n\n  ", "\n"], ["\n  ", ";\n  transition: transform 0.5s ease-in-out;\n  margin-left: 0.3rem;\n\n  ", "\n"])), colors.icon, function (_a) {
    var showMore = _a.showMore;
    return showMore && "transform: rotate(-180deg);";
});
export var ShowMoreContent = styled.div(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  ", "\n  overflow: hidden;\n  /* display: -webkit-box;\n  -webkit-box-orient: vertical; \n -webkit-line-clamp: 3;*/\n  padding-right: 1rem;\n  padding-bottom: 0.3rem;\n\n  ", "\n"], ["\n  ", "\n  overflow: hidden;\n  /* display: -webkit-box;\n  -webkit-box-orient: vertical; \n -webkit-line-clamp: 3;*/\n  padding-right: 1rem;\n  padding-bottom: 0.3rem;\n\n  ",
    "\n"])), function (_a) {
    var showMore = _a.showMore;
    return showMore && "";
}, function (_a) {
    var showMore = _a.showMore;
    return showMore && " overflow: auto; /* -webkit-line-clamp: inherit;*/";
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
