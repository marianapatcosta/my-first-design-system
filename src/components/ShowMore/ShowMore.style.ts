import { MouseEventHandler } from 'react';
import styled from 'styled-components';
import { baseStyles, colors } from '../../styles';

export const StyledShowMore = styled.div`
  ${baseStyles}
`;

export const StyledShowMoreLink = styled.a<{
  onClick?: MouseEventHandler<HTMLAnchorElement>;
}>`
  text-decoration: none;
  color: ${colors.highlight};
  margin: 0.5rem 0;
  float: right;
  cursor: pointer;
`;

export const StyledShowMoreIcon = styled.img<{
  showMore?: boolean;
}>`
  ${colors.iconHighlight};
  transform: rotate(0deg);
  transition: transform 0.5s ease-in-out;
  margin-left: 0.5rem;

  ${({ showMore }) => showMore && `transform: rotate(-180deg);`}
`;

export const ShowMoreContent = styled.div<{ showMore?: boolean }>`
  ${({ showMore }) => showMore && ``}
  overflow: hidden;
  /* display: -webkit-box;
  -webkit-box-orient: vertical; 
 -webkit-line-clamp: 3;*/
  padding-right: 1rem;
  padding-bottom: 0.3rem;

  ${({ showMore }) =>
    showMore && ` overflow: auto; /* -webkit-line-clamp: inherit;*/`}
`;
