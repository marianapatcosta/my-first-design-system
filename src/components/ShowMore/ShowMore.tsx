import React, { CSSProperties } from 'react';
import { StyledShowMore } from './ShowMore.style';

export interface ShowMoreProps {
  /**
   * element id 
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
};

export const ShowMore: React.FC<ShowMoreProps>= ({ ...props}) => {
  return (
    <StyledShowMore  {...props}>
      
    </StyledShowMore>
  );
};


ShowMore.defaultProps = {
  id: '',
  className: '',
  style: {}
};