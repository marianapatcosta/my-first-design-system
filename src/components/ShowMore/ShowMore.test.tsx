import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ShowMore } from './ShowMore';

describe('ShowMore', () => {
  const defaultProps = {
    content: `Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
  Donec vitae mollis velit. Vestibulum lacinia nisi vitae lorem sollicitudin, vel ultricies enim suscipit. 
  Etiam vel sodales dolor. Nunc iaculis quis arcu faucibus tincidunt. Maecenas blandit dapibus porttitor. 
  Quisque sed mauris elit. Ut placerat, augue ut blandit ultricies, nunc nulla vehicula mi, in vehicula urna lectus nec nibh. 
  Aenean quis risus imperdiet, bibendum mauris vitae, consequat tortor. Donec aliquam ante vel turpis fringilla facilisis. 
  Integer varius non lacus id ultrices. In eleifend dolor ante, congue placerat elit maximus eget. Duis varius ligula erat, et varius neque euismod
   ac. In vitae posuere nisl. Sed non justo tortor. Integer risus est, feugiat at ultrices vitae, pretium in est.`,
  };

  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<ShowMore {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should display the first part of the content but not the last part', () => {
    const { getByText } = render(<ShowMore {...defaultProps} />);

    const text = getByText(/Lorem ipsum dolor/i);
    const style = window.getComputedStyle(text);
    expect(style.maxHeight).not.toBe('inherit');
  });

  test('should display the all content on showMore click', () => {
    const { getByText, queryByText } = render(<ShowMore {...defaultProps} />);
    const showMoreLink = getByText(/show more/i);
    showMoreLink && fireEvent.click(showMoreLink);
    expect(queryByText(/Lorem ipsum dolor/i)).toBeTruthy();
    expect(queryByText(/pretium in est/i)).toBeTruthy();
  });

  test('render `show less`label after click', () => {
    const { getByText } = render(<ShowMore {...defaultProps} />);
    const showMoreLink = getByText(/show more/i);
    showMoreLink && fireEvent.click(showMoreLink);
    getByText(/show less/i);
  });
});
