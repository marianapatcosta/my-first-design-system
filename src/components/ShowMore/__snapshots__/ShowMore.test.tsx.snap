// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMore renders without errors and matches snapshot 1`] = `
<div
  class="sc-bdfBwQ jwwsqO"
  id=""
>
  <div
    class="sc-hKgILt aMneb"
    style="max-height: 5.6rem;"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
  Donec vitae mollis velit. Vestibulum lacinia nisi vitae lorem sollicitudin, vel ultricies enim suscipit. 
  Etiam vel sodales dolor. Nunc iaculis quis arcu faucibus tincidunt. Maecenas blandit dapibus porttitor. 
  Quisque sed mauris elit. Ut placerat, augue ut blandit ultricies, nunc nulla vehicula mi, in vehicula urna lectus nec nibh. 
  Aenean quis risus imperdiet, bibendum mauris vitae, consequat tortor. Donec aliquam ante vel turpis fringilla facilisis. 
  Integer varius non lacus id ultrices. In eleifend dolor ante, congue placerat elit maximus eget. Duis varius ligula erat, et varius neque euismod
   ac. In vitae posuere nisl. Sed non justo tortor. Integer risus est, feugiat at ultrices vitae, pretium in est.
  </div>
  <a
    class="sc-gsTCUz hRzuBh"
    data-testid="show-more-link"
    href=""
  >
    show more
     
    <img
      alt="show more icon"
      class="sc-dlfnbm kPXudd"
      src="double-chevron.svg"
    />
  </a>
</div>
`;
